### ISSUE TITLE
[Community Consultation] Should mentoring on side exercises be opt-in per exercise

### ISSUE THREAD

**Issue by iHiD** [exercism/exercism#4305]
### Proposition

Make mentoring on side exercises in mentored mode opt-in per exercise.

### Rational
- We are currently able to keep on top of core exercises, but don't have enough mentors to keep on top of side exercises.
- People are confused/frustrated that they get same-day response to core exercises, but have month-long wait for side exercises.
- We also don't know if everyone wants to feedback on all their side exercises or if they're just practicing. 
- Last week we added a link to community solutions on side exercises to give people some ability to learn without mentoring.

By giving learners who submit an exercise two options to _"Learn from others' solutions"_ or _"Request mentoring (average wait is x days)"_:
- People can only use the mentors if they actually want help
- The total volume of solutions should decrease meaning the mentors actually get to the learners who want help
- Learners expectations are met about wait times etc.

**I would value people's opinions on whether they think this is good or bad idea, if there's something else we should be doing with side exercises instead, and if we have any blind spots that we've not considered.**

Please note, I don't want this to become a conversation on whether core mentoring should be compulsory (it's here to stay in mentored mode), and the final conversation will happen with the Product Team, but we'd like a range of opinions to consider when we make the decision. Thanks! :)

**Comment by F3PiX** [comment]
I'm so glad you asked. :-) The main problem I see with the current system is that it's not very encouraging to do side exercises, and I'm convinced we should encourage it. 

The idea of an extra button is nice! And there could be one on core exercises too, as a 'ready for review' button. Sometimes, as a student I want to try different solutions, and during that process I'd rather postpone the mentor's review. 
Note: AFAIK the solutions pages do not include solutions since mentoring came into live. (I thought I opened an issue for that but I can't find it. Will do that shortly.) Learning from the solutions page may be more effective if the community solutions are filtered in descending order, so that the latest mentoring expertise becomes manifest on the community pages as well. 

On the other hand, from a mentor's point of view, I'd prefer that the early side exercises would get _priority_ over core exercises. For background, in the Ruby Track, the first core exercise discloses some basic side exercises, that are easy to mentor, some even with mostly pre-cooked answers. And those are also great opportunities to discuss some super basic Ruby principles.
I feel that mentoring subsequent core exercises would be both easier and more interesting if those basic principles would have been addressed before the student submits more complex (core) exercises. Because I'd expect that the entry level submission would be more often already idiomatically correct. (Ruby example: so that we don't have to mentor for... loops into Enumerables in exercises that are designed to cover more advanced topics.) 

A different line of thought:
When I saw the latest posting on the queues for each language, the Ruby queue has something around 10 cores in the queue, and far, far more sides (IIRC over 2000). For languages where that's the case, would it be possible to add some side exercises for each core exercise that is queued?
Especially now we learn that core exercises have hardly any waiting time. 
If it's doable to make that distinction between tracks. 

And another line of thought (formerly known as a Wild Idea):
How about queuing not by core or side, but by 'time between last and previous submission', or 'number of submissions in the last week'? In that case, the most active students will get their exercises mentored faster, including side exercises, oldest first. And as a side effect, it may encourage doing side exercises, because the faster mentoring would be an incentive.

**Comment by NobbZ** [comment]
Optional mentoring for side-exercises could potentially decrease the load on mentors, but it is in an area that some tracks are not even able to see yet. So I think tracks mentors wouldn't even notice this change.

The student though might not necessarily understand the implications of getting mentored or getting not mentored and tend to prefer getting mentored, and might choose mentoring anyway.

Nevertheless, if we give students the opportunity to opt-out, we also should give them the opportunity to opt-in later, as we do with independent mode and legacy submissions.

**Comment by cglacet** [comment]
Why not have two score functions, one for core exercises and one for the extra ones, then rank exercise only on their score. The score functions would have age in days as input and score as output. This would avoid having extra exercise lost forever if too many new core exercise keep arriving. 

For example [these two functions](https://www.desmos.com/calculator/lj11evp9vs).

**Comment by kzap** [comment]
I do side exercises waiting to get the main exercises approved.

but it is annoying if the side exercises never get approved.

For me, both for core and side exercises, one should be able to move ahead but still request feedback.

 Basically noone will be able to complete a track because all the side exercises will not get mentored.

I would go with the 2 scores, 1 for core track 100% mentored + and 1 for side exercises completed.

When you submit a solution, the solution page will mark it as Solution Completed and you can opt to request feedback from a mentor and this will put it into the Mentor queue. This way only those who explicitly want feedback for those side exercises and not everyone by default.

or maybe in the future these side exercises could be moved into an advance track so people would also have to do them one by one.

Imho the problem is that there are too many that get compelted at the same time, and thus they fill up the mentor's queue so fast vs the core exercises which is one at a time.

People code faster than code gets approved which is normal since this is not a full time job :)

**Comment by yawpitch** [comment]
The last time I even saw a side exercise in my queue in Python I'd done I think 80-something core ones over 2-3 days, and I never did get to it ... I don't know about other languages, but I feel like the Python track is just completely overwhelmed with just the first 5 core exercises, so yes anything that right-sizes the backlog a bit would help.

I've been pleasantly surprised though to have several side exercises in Rust getting mentors' time while I've got a core item waiting as well, so maybe other tracks are getting a handle on it.

I do very much like the idea @kzap mentioned above of an Advanced track with its own fixed progression as well.

**Comment by axelson** [comment]
My thoughts:
* Make mentoring completely "opt-in" for side exercises
  * So a student can easily move through all side exercises without any mentoring
  * When a student opts-in to mentoring for a side exercise, require them to provide a comment about what they are hoping to get from the mentoring such as ("I want to make this more efficient", or "How can I make my capitalize_word function more elegant")
  * Ideally show a time estimate of how long the side exercise mentoring queue is

Making the side exercises opt-in will decrease the size of the mentoring backlog since not all students will care about mentoring in the side exercises, and showing the time estimate will provide additional back pressure where a student may not request mentoring in a side exercise if they feel that the mentoring backlog for side exercises is already too large.

**Comment by sleepiejohn** [comment]
mine are the same of @axelson, also for me the side exercises are a way for me to practice what I learned from the core ones, so the more I can practice first the better, then if I need I can ask for help from mentors in a particular case.

**Comment by cglacet** [comment]
I just saw the [backlog](https://gist.github.com/iHiD/50f4fc82e63717766ae233f8f3854336) I never though this would be that big, scoring can't even help here. 

Maybe you could add a credit system, where some credits are earned each day and credits can be used to ask for reviews on a given side track exercise? I think it's less frustrating than waiting forever, the "review cost" could be dynamically adjusted depending on the number of pending exercises for this track.

> Add "open for comments from everybody" checkbox 

+1

> More gamification - if somebody got enough points (solved exercises, by mentors, thumbs up for solutions or for good comments) he is actively asked if he doesn't want to be a mentor as he is apparently good enough.

Or you could just allow these users to do code review in the  "open for comments from everybody" section. 

>  I know you want a safe space between student and mentor to avoid beginners being flamed.

That may be something that mentors could decide, when a student seems to be good enough, mentors could mark him/her as "ready to join discussion on side exercises".

**Comment by Emerentius** [comment]
I'm cautiously for making it opt-in.  
Uploading a solution has 2 functions right now:
* Making community solutions visible
* Getting mentoring

And it would be good to separate these. There is little incentive for someone not interested in mentoring to opt-out. Therefore, opt-in is likely better, regardless of which is the more common goal (I suspect it's mentoring).

-------

> @axelson When a student opts-in to mentoring for a side exercise, require them to provide a comment […]

Very much in favor of pushing users to give more comments in general. Just like students don't like getting approved without a word, I'd prefer to have the students say a few words they submit solutions, especially if those are on the more original (or plain weird side). Also when they submit a new solution after getting feedback.

> @yawpitch I've been pleasantly surprised though to have several side exercises in Rust getting mentors' time while I've got a core item waiting as well, so maybe other tracks are getting a handle on it.

Nah, man. We're swamped and racing towards infinity just like Python.

**Comment by yawpitch** [comment]
To infinity, and beyond.

Totally +1 the idea of requiring submit comments; the old V1 CLI had a -c flag that allowed you to submit a comment simultaneous with an iteration, and last I looked at -h it hadn't returned. I miss that, as it saved me having to open up Firefox. Not sure that you couldn't move to requiring a submit message and having an opt-out -C flag instead of an opt-in... a bit of a gentle prompt for the message of not provided might help explain the rationale for the interface change.

**Comment by karelvancamp** [comment]
Would avoid trying to add to many features to tackle this problem (scoring, comments, backlog indication). Just make default to publish non-core solutions directly, optional to request mentor feedback. Later on, more can be added.

**Comment by yilkalargaw** [comment]
Yes I think so. Especially in some language tracks where there are too few mentors it could be discouraging for students to wait on mentoring for non core exercise. I sometimes found it inconvinient to wait on mentoring on non core problems that I had to move to independent mode. I think it should be an opt-in or an opt-out feature for non-core prolems.

**Comment by gDelgado14** [comment]
I have to disagree with @F3PiX on the incentives to do side exercises. 

Whenever I'm waiting on a mentor to review one of my core exercise implementations, I immediately go and do a few side exercises. And once I have a working implementation, I then go and see how other students implemented it. I always look for the exercises with the highest ratings to give me a sense of how well I implemented my own solution. In essence this approach of viewing other people's solutions helps mentor myself.

What I want is to write as much code as possible to gain an intuition for the language. Sometimes the wait times on mentoring impedes this ability. Hence making mentoring optional on side exercises would be great to keep me in a state of "flow".

I am strongly in favor of making side exercises "opt-in".

**Comment by Holland-k** [comment]
I'm not a fan of removing mentoring from the side exercises, whether it is opt-in or otherwise. Really the point of Exercism is the mentoring. Without that, it becomes yet another puzzle site, and no offense, there are a lot of better ones out there. 

I am not sure what the solution is. The tracks I care most about are some of the most overwhelmed ones and agree there is a problem. And, similar to @gDelgado14, I contribute to that problem by working through multiple side exercises while waiting for approval on the core. But taking away the "special sauce" shouldn't be the answer. 

The addition of comments would be great.

**Comment by xarxziux** [comment]
I second the idea that mentoring is what sets Exercism apart from other coding sites and without it there's little incentive to work on the exercises.  From having done other coding sites, one thing that really stands out for Exercism is how light the unit tests are.  Having 20-30 tests at most (many have less) isn't going to cover much.  This is fine if the understanding that the submission is going to get mentored, and the unit tests are just there to cover the most basic points, but is inadequate if we're moving away from that model.  In contrast I've been on sites that run hundreds of tests, including edge case tests, randomised tests, performance tests, etc.  In order to pass those, your code has to be of much better quality than is demanded of the unit tests here.

On the other hand, I've been in the situation more than once where I've spent hours on a submission that passed most of the unit tests but failed a few edge cases.  With the tests being hidden, I was unable to tell what it was failing on and unable to pass the exercise.  It was this frustration that attracted me to Exercism in the first place.

For me, a hybrid solution would be best.  Have lots and lots of unit tests.  Test lots of edge cases, have lots of randomised tests, have some form of performance metric.  Then offer a mentor option if your code gets part of the way but fails the performance metrics or some edge cases.  This would represent the best of both worlds if the "mentor everything" model is struggling, but I appreciate that it would mean a lot of extra work to implement.

**Comment by iHiD** [comment]
OK, thank you all for your comments. Very helpful.

I'm going to go ahead and making mentoring side-exercises optional, with a limit of how many exercises can be submitted for mentoring simultaneously. There's a tracking issue for that here: https://github.com/exercism/exercism/issues/4375
